### PR TITLE
add time in state.

### LINKED ISSUE #916: Insert logged times into state
## 🚀 Feature

By default, Engine measures times for single epoch and total run. These times are logged with `engine.logger`:
https://github.com/pytorch/ignite/blob/efc45a79ee0a5bf8105c3caff194758439db2426/ignite/engine/engine.py#L793-L796
and
https://github.com/pytorch/ignite/blob/efc45a79ee0a5bf8105c3caff194758439db2426/ignite/engine/engine.py#L497-L500

 Idea is to put these times (at least `time_taken` for epoch and total time) into `engine.state` under a name (e.g. "times"). Such that we can also use this information without need to use `Timer` or setup logger.

### PR DESCRIPTION
**Pull request by erip**
Fixes #916

Description: Adds timing info for epoch completion and run completion to the engine's state.

Check list:
* [x] New tests are added (if a new feature is added)
* [x] New doc strings: description and/or example code are in RST format
* [x] Documentation is updated (if required)

I'm feeling a bit sketchy about the tests, so I'd love another pair of eyes on them.

### CODE REVIEW

File: tests/ignite/engine/test_engine.py [hunk 1]
```diff
@@ -366,6 +367,40 @@ def test_state_get_event_attrib_value():
...
+
+        num_examples = len(data)
+
+        engine.run(data, max_epochs=max_epochs, epoch_length=epoch_length)
+
+        lower_bound_time = num_examples / epoch_length / max_epochs
+        upper_bound_time = 1.1 * lower_bound_time
+
+        def init(engine):
+            engine.state.times = {Events.EPOCH_COMPLETED.name: 0.0, Events.COMPLETED.name: 0.0}
```
Comment (vfdev-5) [inline tests/ignite/engine/test_engine.py:383]: I think we should attach this automatically without explicitly require the user

**Comment by vfdev-5** [comment]
@erip I think we should keep previous behavious about logging h,m,s times.

About the tests, I think we can just check the presence of values.

**Comment by erip** [comment]
You want to log the h,m,s times in the engine, but store the timestamp?

**Comment by vfdev-5** [comment]
> You want to log the h,m,s times in the engine, but store the timestamp?

Yes, I want to keep previous engine's behaviour and in addition to store the timestamp.

File: tests/ignite/engine/test_engine.py (hunk 1 above)
Comment (erip) [reply tests/ignite/engine/test_engine.py:383]: Excellent. This initialization happens in the `State` constructor, but I forgot to remove this from the test; that said, I think that's slightly different that wiring up the event handler.

File: ignite/engine/engine.py [hunk 2]
```diff
@@ -433,7 +433,7 @@ def terminate_epoch(self) -> None:
         )
         self.should_terminate_single_epoch = True
 
-    def _run_once_on_dataset(self) -> Tuple[int, int, int]:
+    def _run_once_on_dataset(self) -> float:
```
Comment (vfdev-5) [inline ignite/engine/engine.py:436]: @erip could you please revert this

File: tests/ignite/engine/test_engine.py (hunk 1 above)
Comment (erip) [reply tests/ignite/engine/test_engine.py]: Is there some place where it makes sense to wire this or is `engine.run` good enough?

File: ignite/engine/engine.py [hunk 3]
```diff
@@ -497,10 +497,7 @@ def _run_once_on_dataset(self) -> Tuple[int, int, int]:
             self.logger.error("Current run is terminating due to exception: %s.", str(e))
             self._handle_exception(e)
 
-        time_taken = time.time() - start_time
-        hours, mins, secs = _to_hours_mins_secs(time_taken)
-
-        return hours, mins, secs
+        return time.time() - start_time
```
Comment (vfdev-5) [inline ignite/engine/engine.py:500]: and here, please revert this too. And we are almost done.

File: ignite/engine/engine.py (hunk 2 above)
Comment (erip) [reply ignite/engine/engine.py:436]: If this is reverted, we'll need to unparse the tuple to a timestamp unless we want to put the state-modifying code in the `_run_once_on_dataset`.

Comment (vfdev-5) [reply ignite/engine/engine.py:436]: Yes, you are right ! I didn't get it completely, sorry.

Comment (erip) [reply ignite/engine/engine.py]: Oh, so I should keep it as is with `-> float`? Since it's an internal method it shouldn't break anything (or anyone).

**Comment by vfdev-5** [comment]
Looking into this code now:
```python
                time_taken = self._run_once_on_dataset()
                self.state.times[Events.EPOCH_COMPLETED.name] = time_taken
                hours, mins, secs = _to_hours_mins_secs(time_taken)
                self.logger.info("Epoch[%s] Complete. Time taken: %02d:%02d:%02d", self.state.epoch, hours, mins, secs)
                if self.should_terminate:
                    break
                self._fire_event(Events.EPOCH_COMPLETED)
```
In case if `should_terminate` is True, we log and set the time of a completed epoch, but we do not trigger the event... I need to think if this is OK or we need to trigger the event...

**Comment by vfdev-5** [comment]
> Excellent. This initialization happens in the State constructor, but I forgot to remove this from the test; that said, I think that's slightly different than wiring up the event handler.

> Is there some place where it makes sense to wire this or is engine.run good enough?

Good point, we need to check what happens if we run multiple times evaluator and whether times are still correct. Do we need to include that into state dict (maybe not) ?

**Comment by vfdev-5** [comment]
And @erip I think it is déjàvu, are these values good for a single process and for DDP multi-process ?

**Comment by erip** [comment]
>we need to check what happens if we run multiple times 

Ok, I'll add another test. :)

> Do we need to include that into state dict?

Currently no. It's unclear if these would be useful in the state info. If they were, I imagine it would just be simple enough to (de)serialize the values.

File: ignite/engine/events.py [hunk 4]
```diff
@@ -270,6 +270,7 @@ def __init__(self, **kwargs):
         self.metrics = {}
         self.dataloader = None
         self.seed = None
+        self.times = {Events.EPOCH_COMPLETED.name: 0.0, Events.COMPLETED.name: 0.0}
```
Comment (vfdev-5) [inline ignite/engine/events.py:273]: Maybe, we need to init values as `None`, such that if, for example, user request values of `Events.COMPLETED.name` it wont say that run took 0.0 seconds... Thoughts ?

File: tests/ignite/engine/test_engine.py [hunk 5]
```diff
@@ -366,6 +367,34 @@ def test_state_get_event_attrib_value():
...
+        num_examples = len(data)
+
+        engine.run(data, max_epochs=max_epochs, epoch_length=epoch_length)
+
+        lower_bound_time = num_examples / epoch_length / max_epochs
+        upper_bound_time = 1.1 * lower_bound_time
+
+        def check_epoch_time(engine):
+            # TODO: can we make this test more meaningful?
+            assert engine.state.times[Events.EPOCH_COMPLETED.name] != 0
```
Comment (vfdev-5) [inline tests/ignite/engine/test_engine.py:384]:
```suggestion
            assert engine.state.times[Events.EPOCH_COMPLETED.name] >= 0.01 * epoch_length 
```
something like that ?

File: tests/ignite/engine/test_engine.py [hunk 6]
```diff
@@ -366,6 +367,34 @@ def test_state_get_event_attrib_value():
...
+def test_time_stored_in_state():
+    def _test(data, max_epochs, epoch_length):
+
+        engine = Engine(lambda e, b: time.sleep(0.01))
+
+        num_examples = len(data)
+
+        engine.run(data, max_epochs=max_epochs, epoch_length=epoch_length)
+
+        lower_bound_time = num_examples / epoch_length / max_epochs
```
Comment (vfdev-5) [inline tests/ignite/engine/test_engine.py:379]: these two lines are unused

File: tests/ignite/engine/test_engine.py [hunk 7]
```diff
@@ -366,6 +367,34 @@ def test_state_get_event_attrib_value():
...
+        upper_bound_time = 1.1 * lower_bound_time
+
+        def check_epoch_time(engine):
+            # TODO: can we make this test more meaningful?
+            assert engine.state.times[Events.EPOCH_COMPLETED.name] != 0
+
+        def check_completed_time(engine):
+            # TODO: can we make this test more meaningful?
+            assert engine.state.times[Events.COMPLETED.name] != 0
```
Comment (vfdev-5) [inline tests/ignite/engine/test_engine.py:388]:
```suggestion
            assert engine.state.times[Events.COMPLETED.name] >= 0.01 * epoch_length * max_epochs
```

File: ignite/engine/events.py (hunk 4 above)
Comment (erip) [reply ignite/engine/events.py:273]: I agree. This makes the wiring more intentional, too.

File: tests/ignite/engine/test_engine.py (hunk 6 above)
Comment (erip) [reply tests/ignite/engine/test_engine.py:379]: They are? I thought the top line was driving the attached handlers which make assertions.

**Comment by vfdev-5** [comment]
@erip thanks for the updates! Could you please update the docstring of `Engine` and show the usage in `examples/mnist/mnist.py`.

**Comment by erip** [comment]
Absolutely

**Comment by erip** [comment]
@vfdev-5 thanks for catching the missed TODO comments. Please let me know if there's anything else you want me to polish

**Comment by vfdev-5** [comment]
@erip I thinks we are good with this one. Could you, please, take a look at this issue : https://github.com/pytorch/ignite/issues/784
 with `Frequency` you worked previously ?

**Comment by erip** [comment]
Are the tests normally this flaky? The failure seems unrelated, but I've seen the same test fail a few times.

**Comment by vfdev-5** [comment]
@erip OK, I understand now why tests are failing. In FastaiLRFiner tests we check that there is no more attached handlers, by default `Engine` has no internally attached handlers
```
    def test_with_attach(lr_finder, to_save, dummy_engine, dataloader):

        with lr_finder.attach(dummy_engine, to_save=to_save) as trainer_with_finder:
            trainer_with_finder.run(dataloader)

        assert lr_finder.get_results() is not None
        for event in dummy_engine._event_handlers:
>           assert len(dummy_engine._event_handlers[event]) == 0
```
and here we added `_init_timers` on started. Maybe we can call this method inside `run` and keep previous behaviour.

**Pull request merged**